internal/handler/cmd_query.go

```diff
@@ -58,3 +84,65 @@ func (h *Handler) CmdQuery(ctx context.Context, query *wire.OpQuery) (*wire.OpRe
 		"OpQuery: "+cmd,
 	)
 }
+
+// speculativeAuthenticate uses db and mechanism to authenticate and returns the document
+// to assign for op query speculativeAuthenticate response field if the authentication is successful.
+func (h *Handler) speculativeAuthenticate(ctx context.Context, document *types.Document) (*types.Document, error) {
+	dbName, err := common.GetRequiredParam[string](document, "db")
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	mechanism, err := common.GetRequiredParam[string](document, "mechanism")
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	switch mechanism {
+	case "PLAIN":
```

Comment:
More tests will be covered once #1877 is merged.

Reply:
>Where did you get this information from?

By seeing `task run-proxy` and how this PR behaves on `task run` with successful authentication. I tried it by `task mongosh` then running `db.auth({user:'user',pwd:'pwd',mechanism: 'SCRAM-SHA-256'})`.

When speculative authentication is not implemented, client makes `isMaster`, `saslStart` then `saslContinue`. Once speculative authenticate is implemented and the speculative authenticate succeeds with `isMaster` then `saslContinue` is called.

Okay skip might not be the right term but what I'm emphasizing is that the client does not call `saslStart` command once speculative authenticate has succeeded. So one round trip of saslStart command is skipped. Am I understanding this wrong?

> Note I am connecting to MongoDB here, it uses an OP_QUERY when connecting to FerretDB so I don't think we need to handle the OP_MSG case.

Good, let's just handle it in op_query of isMaster and possibly hello.

> The hello and isMaster responses are nearly identical, so I think either will work given that we continue the SCRAM conversation.

Yes they are similar but the difference is `"ismaster", true` field is set on `isMaster` response. Does this have an impact :thinking:

Also I'm not sure how to test speculative authenticate works with `hello`, but it seems that you have `hello` op query with speculative authenticate field. Could you test my PR it works with `hello` the way you did that above?

Below is the request for speculative authenticate captured by FerretDB. See how it's `isMaster` op query.
```
{                                                                                                                                                                                                           
  "Flags": 0,                                                                                                                                                                                               
  "FullCollectionName": "admin.$cmd",                                                                                                                                                                       
  "NumberToReturn": -1,                                                                                                                                                                                     
  "NumberToSkip": 0,                                                                                                                                                                                        
  "Query": {                                                                                                                                                                                                
    "$k": [                                                                                                                                                                                                 
      "ismaster",                                                                                                                                                                                           
      "helloOk",                                                                                                                                                                                            
      "client",                                                                                                                                                                                             
      "compression",                                                                                                                                                                                        
      "speculativeAuthenticate"                                                                                                                                                                             
    ],                                                                          
    "ismaster": 1,                      
    "helloOk": true,                                                            
    "client": {                                                                 
      "$k": [                                                                   
        "application",                                                          
        "driver",                                                               
        "platform",                                                             
        "os"                                                                    
      ],                                                                        
      "application": {                                                          
        "$k": [                                                                 
          "name"                                                                
        ],                                                                      
        "name": "mongosh 2.1.4"                                                 
      },                                                                        
      "driver": {                                                               
        "$k": [                                                                 
          "name",                                                               
          "version"                                                             
        ],                                                                      
        "name": "nodejs|mongosh",                                               
        "version": "6.3.0|2.1.4"                                                
      },                                                                        
      "platform": "Node.js v20.11.0, LE",                                       
      "os": {                                                                   
        "$k": [                                                                 
          "name",                                                               
          "architecture",                                                       
          "version",                                                            
          "type"                                   
        ],                                         
        "name": "linux",                           
        "architecture": "x64",                     
        "version": "5.15.49-linuxkit-pr",                                                             
        "type": "Linux"                            
      }                                            
    },                                             
    "compression": [                               
      "none"                                       
    ],                                             
    "speculativeAuthenticate": {                   
      "$k": [                                      
        "saslStart",                               
        "mechanism",                               
        "payload",                                 
        "autoAuthorize",                           
        "options",                                 
        "db"                                       
      ],                                           
      "saslStart": 1,                              
      "mechanism": "SCRAM-SHA-256",                
      "payload": {                                 
        "$b": "biwsbj11c2VyMSxyPUNCUnJLelJQRy9qOEY2UUxlcUZwVDFDdGxuaGtzc0xI",                         
        "s": 0                                     
      },                                           
      "autoAuthorize": 1,                          
      "options": {                                 
        "$k": [                                    
          "skipEmptyExchange"                      
        ],                                         
        "skipEmptyExchange": true                  
      },                                           
      "db": "test"                                 
    }                                              
  }                                                
}                                                  
```

Reply:
> Okay skip might not be the right term but what I'm emphasizing is that the client does not call `saslStart` command once speculative authenticate has succeeded.

Ah I see you meant, and your understanding is correct. It just avoids a round-trip by embedding the `saslStart` command in the initial handshake.

The above request is something referred as helloOk Protocol Negotation and is used to check that the server supports the `hello` command because it's a newish command. If the server responds back with `helloOk: true` as well then both parties can start sending `hello` instead of `isMaster` on the connection.

IMO if there's a field with `helloOk` or if the command is a `hello` then we should probably respond with a proper `hello` response, but I also used the `isMaster` [response](https://github.com/FerretDB/FerretDB/pull/3904/files#diff-87d3df5f25dccdd7b09cc9e6bf04964eff5a2882f82f6c3fcdef423ed27d8f41R46) in my research prototype and it worked.

Reply:
> IMO if there's a field with helloOk or if the command is a hello then we should probably respond with a proper hello response, but I also used the isMaster [response](https://github.com/FerretDB/FerretDB/pull/3904/files#diff-87d3df5f25dccdd7b09cc9e6bf04964eff5a2882f82f6c3fcdef423ed27d8f41R46) in my research prototype and it worked.

Yes I noticed your example has `isMaster` response for both `isMaster` and `hello`. If that's expected, all is good. I just wanted to ensure, because I didn't test it myself :hugs:

Reply:
@chilagrow I checked out the PR and tested it with multiple drivers and it works as intended.

The only issue is with the legacy shell using an OP_MSG, but I think we can handle that in a separate PR.

Reply:
Thanks a lot @b1ron :hugs: . Sure we can handle in separate PR _if_ we want to support that for legacy shell. By the way, I created an issue for hello ok https://github.com/FerretDB/FerretDB/issues/4132